airbyte-integrations/connectors/source-snowflake/src/main/java/io.airbyte.integrations.source.snowflake/SnowflakeDataSourceUtils.java

```diff
@@ -0,0 +1,168 @@
+/*
+ * Copyright (c) 2021 Airbyte, Inc., all rights reserved.
+ */
+
+package io.airbyte.integrations.source.snowflake;
+
+import static java.util.stream.Collectors.joining;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.zaxxer.hikari.HikariConfig;
+import com.zaxxer.hikari.HikariDataSource;
+import io.airbyte.commons.json.Jsons;
+import java.io.IOException;
+import java.net.URI;
+import java.net.URLEncoder;
+import java.net.http.HttpClient;
+import java.net.http.HttpRequest;
+import java.net.http.HttpRequest.BodyPublisher;
+import java.net.http.HttpRequest.BodyPublishers;
+import java.net.http.HttpResponse;
+import java.nio.charset.StandardCharsets;
+import java.time.Duration;
+import java.util.Base64;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Properties;
+import java.util.concurrent.TimeUnit;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class SnowflakeDataSourceUtils {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(SnowflakeDataSourceUtils.class);
+  private static final int PAUSE_BETWEEN_TOKEN_REFRESH_MIN = 7; // snowflake access token's TTL is 10min and can't be modified
+  private static final String REFRESH_TOKEN_URL = "https://%s/oauth/token-request";
+  private static final HttpClient httpClient = HttpClient.newBuilder()
+      .version(HttpClient.Version.HTTP_2)
+      .connectTimeout(Duration.ofSeconds(10))
+      .build();
+
+  /**
+   * Snowflake OAuth access token expires in 10 minutes. For the cases when sync duration is more than
+   * 10 min, it requires updating 'token' property after the start of connection pool.
+   * HikariDataSource brings support for this requirement.
+   *
+   * @param config source config JSON
+   * @return datasource
+   */
+  public static HikariDataSource createDataSource(final JsonNode config) {
+    HikariConfig hikariConfig = new HikariConfig();
+    final StringBuilder jdbcUrl = new StringBuilder(
+        String.format("jdbc:snowflake://%s/?", config.get("host").asText()));
+    jdbcUrl.append(String.format(
+        "role=%s&warehouse=%s&database=%s&schema=%s&JDBC_QUERY_RESULT_FORMAT=%s&CLIENT_SESSION_KEEP_ALIVE=%s",
+        config.get("role").asText(),
+        config.get("warehouse").asText(),
+        config.get("database").asText(),
+        config.get("schema").asText(),
+        // Needed for JDK17 - see
+        // https://stackoverflow.com/questions/67409650/snowflake-jdbc-driver-internal-error-fail-to-retrieve-row-count-for-first-arrow
+        "JSON",
+        true));
+    if (config.has("jdbc_url_params")) {
+      jdbcUrl.append(config.get("jdbc_url_params").asText());
+    }
+    hikariConfig.setJdbcUrl(jdbcUrl.toString());
```

Comment:
can you move the logic of the URL building into a separate method? also, it would be great to cover it with some unit tests

Reply:
done